### PR TITLE
DL: Trap unsupported options for compile and fit params

### PR DESCRIPTION
**Pull request by orhankislal**
JIRA: MADLIB-1309

This commit adds error messages for the unspoorted options for compile
and test params.

Co-authored-by: Jingyi Mei <jmei@pivotal.io>

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/840/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/842/

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in [hunk 1]
```diff
@@ -199,7 +215,34 @@ def get_optimizers():
...
-                  loss=compile_dict['loss'],
-                  metrics=compile_dict['metrics'])
+    compile_dict['optimizer'] = optimizers[opt_name](**final_args) if final_args else opt_name
+    validate_compile_params(compile_dict)
+    model.compile(**compile_dict)
+
+def validate_compile_params(compile_dict):
+    _assert(compile_dict['metrics'] is None or type(compile_dict['metrics']) is list,
+            "wrong input type for compile parameter metrics: multi-output model"
+            "is not supported yet, please pass a list")
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in]: 1. I am confused by the `multi-output model` error message. Can you explain?
1. Should both of these conditions throw the same error message ? This also applies to all the other asserts in this function
1. Keras allows user to pass in a custom metric but we can't support it. We should add this to the docs

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in [hunk 2]
```diff
@@ -199,7 +215,34 @@ def get_optimizers():
...
+    _assert(compile_dict['sample_weight_mode'] is None or
+            compile_dict['sample_weight_mode'] == "temporal",
+            """compile parameter sample_weight_mode can only be "temporal" or None {0}
+            """.format(compile_dict['sample_weight_mode']))
+
+    accepted_compile_params = ['optimizer', 'loss', 'metrics', 'loss_weights',
+                               'weighted_metrics', 'sample_weight_mode']
+    for ckey in compile_dict.keys():
+        _assert(ckey in accepted_compile_params,
+                "{0} is not accepted as a compile parameter yet.".format(ckey))
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in]: Maybe we should also prompt the user to review the dl docs

File: src/ports/postgres/modules/deep_learning/test/unit_tests/test_madlib_keras.py_in [hunk 3]
```diff
@@ -178,11 +178,19 @@ class MadlibKerasFitTestCase(unittest.TestCase):
     def test_split_and_strip(self):
         self.assertEqual(('a','b'), self.subject.split_and_strip(' a = b '))
 
-    def test_parse_fit_params(self):
+    def test_parse_and_validate_fit_params(self):
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/test/unit_tests/test_madlib_keras.py_in]: We should also add negative test cases for `parse_and_validate_fit_params` and `parse_compile_params`. Maybe not now but in a future PR

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in (hunk 1 above)
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in]: Keras accepts a dict as `metrics` if you are building a multi-output model (that we don't support yet).

Comment (kaknikhil) [reply src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in]: I see thanks.
But what if the user passes in just a string? The check will fail and we will print the wrong error message

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in [hunk 4]
```diff
@@ -120,9 +121,14 @@ def parse_compile_params(str_of_args):
...
     opt_name,opt_args = parse_optimizer(compile_dict)
     compile_dict['loss'] = parse_loss(compile_dict)
+    accepted_compile_params = ['metrics', 'loss_weights', 'weighted_metrics', 'sample_weight_mode']
 
-    compile_dict['metrics'] = ast.literal_eval(compile_dict['metrics']) if 'metrics' in compile_dict.keys() else None
-    compile_dict['loss_weights'] = ast.literal_eval(compile_dict['loss_weights']) if 'loss_weights' in compile_dict.keys() else None
+    for key in accepted_compile_params:
+        try:
+            compile_dict[key] = ast.literal_eval(compile_dict[key]) if key in compile_dict.keys() else None
+        except ValueError:
```
[the remark this answers is not in the document]
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in]: Renamed and refactored the functions to make it a bit more consistent.

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in (hunk 1 above)
Comment (orhankislal) [reply src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in]: The keras documentation indicates that it should be a list. Taken from https://keras.io/models/model/
```
Typically you will use metrics=['accuracy']
```

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/856/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/857/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/858/
